### PR TITLE
build: Use C++17 standard

### PR DESCRIPTION
**Pull request by ncesario-lunarg**
@juan-lunarg showed that we are already using C++14 features, so making the minimum explicit.

### CODE REVIEW

**Comment by juan-lunarg** [comment]
> CI Vulkan-ValidationLayers build # 9100 passed.

Probably a good idea to update build.gn and the android build as well

**Comment by ncesario-lunarg** [comment]
> > CI Vulkan-ValidationLayers build # 9100 passed.
> 
> Probably a good idea to update build.gn and the android build as well

I don't see where to make this change, but that is probably just due to my ignorance with GN.

**Comment by juan-lunarg** [comment]
> I don't see where to make this change, but that is probably just due to my ignorance with GN.

Yeah, I was a bit confused by that as well.

**Comment by juan-lunarg** [comment]
The GN build has a minor issue:
```
-std=c++17 -Wno-undefined-bool-conversion -Wno-tautological-undefined-compare -std=c++14
```

Looks like -std=c++14 is overriding the -std=c++17 you set

File: BUILD.gn [hunk 1]
```diff
@@ -86,6 +86,13 @@ config("vulkan_internal_config") {
   if (is_clang) {
     cflags += [ "-Wno-extra-semi" ]
   }
+
+  # TODO is there a more "standard" way to set this?
+  if (!is_win) {
+    cflags += [ "-std=c++17" ]
+  } else {
+    cflags += [ "/std:c++17" ]
+  }
```
Comment (juan-lunarg) [inline BUILD.gn:95]: I found an example for setting the C++ standard:
https://fuchsia.googlesource.com/fuchsia/+/refs/heads/main/build/config/BUILD.gn

Looks like there is a kinda standard approach. Couldn't find anything in the docs.

```
import("//build_overrides/vulkan_validation_layers.gni")

config("language") {
  if (!is_win) {
    cflags_cc += [ "-std=c++17" ]
  } else {
    cflags_cc += [ "/std:c++17" ]
  }
}
```

Comment (juan-lunarg) [reply BUILD.gn:95]: Note you want to use the `cflags_cc` variable otherwise you'll affect the compilation of `xxhash` which is a c library

Comment (juan-lunarg) [reply BUILD.gn:95]: Hmm, now I'm not so sure after looking at this chromium build.
https://github.com/timniederhausen/gn-build/blob/923683700a39a53ecf467974cbf29f3405358292/config/BUILD.gn#L807

Comment (juan-lunarg) [reply BUILD.gn:95]: Looks like the chromium builds don't have  a standardized approach like `CXX_STANDARD` in CMake. So each repo makes up their own approach.

Comment (ncesario-lunarg) [reply BUILD.gn:95]: :+1: thanks! I can cherry pick the changes you made in #4791 here.

The internal build failure looks spurious.

Comment (juan-lunarg) [reply BUILD.gn:95]: I wouldn't cherry pick yet since my change failed to build.

Comment (juan-lunarg) [reply BUILD.gn]: Latest version of my PR appears to pass.

File: BUILD.md [hunk 2]
```diff
@@ -531,7 +531,7 @@ layer validation tests use the "googletest" testing framework.)
 
 This repository has been built and tested on the two most recent Ubuntu LTS
 versions. Currently, the oldest supported version is Ubuntu 16.04, meaning
-that the minimum officially supported C++11 compiler version is GCC 5.4.0,
+that the minimum officially supported C++17 compiler version is GCC 5.4.0,
```
Comment (juan-lunarg) [inline BUILD.md:534]: The Ubuntu version and GCC version don't seem correct.

**Review by juan-lunarg — CHANGES_REQUESTED** [review]

File: BUILD.md (hunk 2 above)
Comment (juan-lunarg) [inline BUILD.md:534]: Min Ubuntu version should be 18.04, not sure what the min GCC version should be.

**Review by juan-lunarg — APPROVED** [review]

**Pull request merged**